File: oracle-js/canon.js
export const CANON_SPEC_V1 = "asx://canon/json.bytes.v1";

function sortObject(value) {
  if (Array.isArray(value)) {
    return value.map(sortObject);
  }
  if (value && typeof value === "object" && value.constructor === Object) {
    const sorted = {};
    Object.keys(value)
      .sort()
      .forEach((key) => {
        sorted[key] = sortObject(value[key]);
      });
    return sorted;
  }
  return value;
}

export function canonJsonBytesV1(obj) {
  const sorted = sortObject(obj);
  const json = JSON.stringify(sorted);
  return new TextEncoder().encode(json);
}

File: python/oracle/canon.py
from __future__ import annotations

import json
from typing import Any


CANON_SPEC_V1 = "asx://canon/json.bytes.v1"


def canon_json_bytes_v1(obj: Any) -> bytes:
    """Return canonical JSON bytes for ABI hashing.

    Canonicalization rules (v1):
    - UTF-8 encoding
    - Object keys sorted lexicographically
    - No insignificant whitespace
    """
    canon = json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    return canon.encode("utf-8")
